README.md
# flutter_overboard  
[![pub package](https://img.shields.io/badge/pub-v3.1.1-blue)](https://pub.dev/packages/flutter_overboard)  


Onboarding widget for flutter to create beautiful onboarding slides with minimal code.


## Demo   
 ![Example Gif](https://github.com/kumar-aakash86/flutter_overboard/blob/master/screenshots/example.gif)


## Usage
Add following command in your **pubspec.yaml** & install package

`flutter_overboard:2.0.1`
    


_**Import in your dart page**_
```
import 'package:flutter_overboard/flutter_overboard.dart';
```  
  
  
**_Create a pages array like_**   
```
    final pages = [
        new PageModel(
            color: const Color(0xFF0097A7),
            imageAssetPath: 'assets/01.png',
            title: 'Screen 1',
            body: 'Share your ideas with the team',
            doAnimateImage: true),
        new PageModel(
            color: const Color(0xFF536DFE),
            imageAssetPath: 'assets/02.png',
            title: 'Screen 2',
            body: 'See the increase in productivity & output',
            doAnimateImage: true),
        new PageModel(
            color: const Color(0xFF9B90BC),
            imageAssetPath: 'assets/03.png',
            title: 'Screen 3',
            body: 'Connect with the people from different places',
            doAnimateImage: true),
    ];
```   
  
**_You can also pass widgets as page model_** 
```
    PageModel.withChild(
        child: new Padding(
          padding: new EdgeInsets.only(bottom: 25.0),
          child: new Image.asset('assets/02.png', width: 300.0, height: 300.0),
        ),
        color: const Color(0xFF5886d6),
        doAnimateChild: true)
```
  
  
**_Add follwing in you dart code widget_**
```
    OverBoard(
        pages: pages,
        showBullets: true,
        skipCallback: () {
          // WRITE SKIP BUTTON ACTION HERE
        },
        finishCallback: () {
          // WRITE THE FINISH BUTTON ACTION HERE
        },
      ),
```
  
**That's it. You are done with the setup now try to run your app.**

  
**_To customize the circle reveal center point use (Overboard Widget)_**
```
    center: Offset(dx, dy);
```
  
  
**_To customize the text of buttons (Overboard Widget)_**
```
    skipText: "Go Out",
    nextText: "Go Forward",
    finishText: "END",
```
  
  
**_To customize the color of buttons (Overboard Widget)_**
```
    buttonColor: Colors.blue,
``` 

**_To customize the color of bullets (Overboard Widget)_**
```
    activeBulletColor: Colors.white,
    inactiveBulletColor: Colors.white30,
```

**_To customize the background of pages (Overboard Widget)_**
```
  backgroundProvider: NetworkImage('https://picsum.photos/720/1280')
```
OR
```
  backgroundProvider: ImageProvider('assets/images/bg.jpg')
```

**_To customize the color of page text (PageModel Widget)_**
```
    titleColor: Colors.blue,
    bodyColor: Colors.red,
```

**_To add the scroll support using keyboard & mouse wheel (Overboard Widget)_**
```
    allowScroll: true,
``` 
     
### Overboard Widget
| PROPERTY            | TYPE            | REQUIRED | DETAILS                                        |
| ------------------- | --------------- | -------- | ---------------------------------------------- |
| pages               | List<PageModel> | yes      | List of pages to render on-boarding            |
| center              | Offset          | no       | Offset to set center point of revealing circle |
| showBullets         | Boolean         | no       | Enable/disable bullets visibility              |
| skipText            | String          | no       | Customize skip button text                     |
| nextText            | String          | no       | Customize next button text                     |
| finishText          | String          | no       | Customize finish button text                   |
| buttonColor         | Color           | no       | Customize button color                         |
| activeBulletColor   | Color           | no       | Customize active bullet color                  |
| inactiveBulletColor | Color           | no       | Customize inactive bullet color                |
| backgroundProvider  | ImageProvider   | no       | Overboard background image provider            |
| skipCallback        | VoidCallback    | no       | Skip button click callback                     |
| finishCallback      | VoidCallback    | no       | Finish button click callback                   |


### PageModel Widget
| PROPERTY       | TYPE    | REQUIRED | DETAILS                                |
| -------------- | ------- | -------- | -------------------------------------- |
| color          | Color   | no       | Background color of the page           |
| imageAssetPath | String  | no       | Image path from asset to show in page  |
| title          | String  | no       | Title text of the page                 |
| body           | String  | no       | Body text of the page                  |
| child          | Widget  | no       | Custom widget to pass as image in page |
| doAnimateChild | Boolean | yes      | To enable/disable child animation      |
| doAnimateImage | Boolean | yes      | To enable/disable image animation      |
| titleColor     | Color   | no       | Change color of title text             |
| bodyColor      | Color   | no       | Change color of body text              |


## Example
```
import 'package:flutter/material.dart';

import 'package:flutter_overboard/flutter_overboard.dart';

void main() => runApp(MyApp());

class MyApp extends StatelessWidget {
  @override
  Widget build(BuildContext context) {
    return MaterialApp(
      title: 'Flutter Overboard Demo',
      debugShowCheckedModeBanner: false,
      theme: ThemeData(
        primarySwatch: Colors.blue,
      ),
      home: MyHomePage(title: 'Flutter Demo Home Page'),
    );
  }
}

class MyHomePage extends StatefulWidget {
  MyHomePage({Key key, this.title}) : super(key: key);

  final String title;

  @override
  _MyHomePageState createState() => _MyHomePageState();
}

class _MyHomePageState extends State<MyHomePage> {
  final GlobalKey<ScaffoldState> _globalKey = new GlobalKey<ScaffoldState>();

  @override
  Widget build(BuildContext context) {
    return Scaffold(
      key: _globalKey,
      body: OverBoard(
        pages: pages,
        showBullets: true,
        skipCallback: () {
          _globalKey.currentState.showSnackBar(SnackBar(
            content: Text("Skip clicked"),
          ));
        },
        finishCallback: () {
          _globalKey.currentState.showSnackBar(SnackBar(
            content: Text("Finish clicked"),
          ));
        },
      ),
    );
  }

  final pages = [
        PageModel(
            color: const Color(0xFF0097A7),
            imageAssetPath: 'assets/01.png',
            title: 'Screen 1',
            body: 'Share your ideas with the team',
            doAnimateImage: true),
        PageModel(
            color: const Color(0xFF536DFE),
            imageAssetPath: 'assets/02.png',
            title: 'Screen 2',
            body: 'See the increase in productivity & output',
            doAnimateImage: true),
        PageModel(
            color: const Color(0xFF9B90BC),
            imageAssetPath: 'assets/03.png',
            title: 'Screen 3',
            body: 'Connect with the people from different places',
            doAnimateImage: true),
        PageModel.withChild(
            child: Padding(
              padding: EdgeInsets.only(bottom: 25.0),
              child: Image.asset('assets/02.png', width: 300.0, height: 300.0),
            ),
            color: const Color(0xFF5886d6),
            doAnimateChild: true)
  ];
}
```


# LICENSE
[MIT LICENSE](https://github.com/kumar-aakash86/flutter_overboard/blob/master/LICENSE.md)


CHANGELOG.md
## [3.1.1] - 18th Jan 2022
* Keyboard & mouse scroll navigation

## [3.0.0] - 30th Sept, 2021
* Migrated to AndroidX

## [2.0.1] - 13th April, 2021
* Null safety added
* Button text color support

## [1.1.5] - 20th November, 2019
* Custom text for skip, next & finish buttons

## [1.1.4] - 18th September, 2019
* Added skip button callback
* Circle ripple center offset property

## [1.1.0-1.1.3] - 29th August, 2019
* Option to pass complete widget with "PageModel.withChild"
* Enable animation for child widget
* Enable animation for page widget images

## [1.0.5] - 29th August, 2019
* Handle UI with many items

## [1.0.1-1.0.4] - 28th August, 2019
* Minor fixes

## [1.0.0] - 28th August, 2019

* Initial release


test/flutter_overboarding_test.dart
// import 'package:flutter_test/flutter_test.dart';

// import 'package:flutter_overboard/flutter_overboard.dart';

void main() {
  // test('adds one to input values', () {
  //   final calculator = Calculator();
  //   expect(calculator.addOne(2), 3);
  //   expect(calculator.addOne(-7), -6);
  //   expect(calculator.addOne(0), 1);
  //   expect(() => calculator.addOne(null), throwsNoSuchMethodError);
  // });
}


lib/src/overboard.dart
import 'dart:math';

import 'package:flutter/gestures.dart';
import 'package:flutter/material.dart';
import 'package:flutter/services.dart';
import 'package:flutter_overboard/src/circular_clipper.dart';
import 'package:flutter_overboard/src/overboard_animator.dart';
import 'package:flutter_overboard/src/page_model.dart';

enum SwipeDirection { LEFT_TO_RIGHT, RIGHT_TO_LEFT, SKIP_TO_LAST }

class OverBoard extends StatefulWidget {
  /// List of pages to render on-boarding
  final List<PageModel> pages;

  /// Offset to set center point of revealing circle
  final Offset? center;

  /// Enable/disable bullets visibility
  final bool showBullets;

  /// Callback method to capture finish button click
  final VoidCallback finishCallback;

  /// Callback method to capture skip button click
  final VoidCallback? skipCallback;

  /// Customize skip button text
  final String? skipText;

  /// Customize next button text
  final String? nextText;

  /// Customize finish button text
  final String? finishText;

  /// Change action button colors
  final Color buttonColor;
  final bool allowScroll;

  /// Change active bullet color
  final Color activeBulletColor;

  /// Change inactive bullet color
  final Color inactiveBulletColor;

  // Overboard background provider
  final ImageProvider<Object>? backgroundProvider;

  OverBoard(
      {Key? key,
      required this.pages,
      this.center,
      this.showBullets = true,
      this.skipText,
      this.nextText,
      this.finishText,
      required this.finishCallback,
      this.skipCallback,
      this.buttonColor = Colors.white,
      this.activeBulletColor = Colors.white,
      this.inactiveBulletColor = Colors.white30,
      this.backgroundProvider,
      this.allowScroll = false})
      : super(key: key);

  @override
  _OverBoardState createState() => _OverBoardState();
}

class _OverBoardState extends State<OverBoard> with TickerProviderStateMixin {
  late OverBoardAnimator _animator;

  ScrollController _scrollController = new ScrollController();
  double _bulletPadding = 5.0, _bulletSize = 10.0, _bulletContainerWidth = 0;

  int _counter = 0, _last = 0;
  int _total = 0;
  double initial = 0, distance = 0;
  Random random = new Random();
  SwipeDirection _swipeDirection = SwipeDirection.RIGHT_TO_LEFT;

  BoxDecoration _boxDecoration = BoxDecoration();
  final scrollDuration = Duration(milliseconds: 1500);
  bool isScrolling = false;

  @override
  void initState() {
    super.initState();

    _animator = new OverBoardAnimator(this);
    _total = widget.pages.length;

    if (widget.backgroundProvider != null)
      _boxDecoration = BoxDecoration(
          image: DecorationImage(image: widget.backgroundProvider!));
  }

  @override
  Widget build(BuildContext context) {
    return Container(
      child: widget.allowScroll ? _keyboardWrapper() : _getStack(),
    );
  }

  _keyboardWrapper() {
    return RawKeyboardListener(
      autofocus: true,
      focusNode: FocusNode(),
      onKey: (RawKeyEvent e) {
        if (e is RawKeyDownEvent) {
          if (e.isKeyPressed(LogicalKeyboardKey.arrowRight) &&
              _counter < _total - 1) {
            _next();
          } else if (e.isKeyPressed(LogicalKeyboardKey.arrowLeft) &&
              _counter > 0) {
            _goBack();
          }
        }
      },
      child: _scrollWrapper(),
    );
  }

  _scrollWrapper() {
    return Listener(
      onPointerSignal: (event) async {
        if (event is PointerScrollEvent) {
          num scrollDelta = event.scrollDelta.dy;
          if (scrollDelta == 0) {
            scrollDelta = event.scrollDelta.dx;
          }
          if (!isScrolling) {
            isScrolling = true;
            if (scrollDelta.isNegative && _counter > 0) {
              _goBack();
            } else if (!scrollDelta.isNegative && _counter < _total - 1) {
              _next();
            }

            await Future.delayed(scrollDuration);
            isScrolling = false;
          }
        }
      },
      child: _getStack(),
    );
  }

  _getStack() {
    return GestureDetector(
      onPanStart: (DragStartDetails details) {
        initial = details.globalPosition.dx;
      },
      onPanUpdate: (DragUpdateDetails details) {
        distance = details.globalPosition.dx - initial;
      },
      onPanEnd: (DragEndDetails details) {
        initial = 0.0;
        setState(() {
          _last = _counter;
        });
        if (distance > 1 && _counter > 0) {
          _goBack();
        } else if (distance < 0 && _counter < _total - 1) {
          _next();
        }
      },
      child: Stack(
        children: <Widget>[
          _getPage(_last),
          AnimatedBuilder(
            animation: _animator.getAnimator(),
            builder: (context, child) {
              return ClipOval(
                  clipper: CircularClipper(
                      _animator.getAnimator().value, widget.center),
                  child: _getPage(_counter));
            },
            child: Container(),
          ),
          Positioned(
            left: 0,
            right: 0,
            bottom: 0,
            child: Row(
              mainAxisAlignment: MainAxisAlignment.spaceEvenly,
              children: <Widget>[
                Opacity(
                  child: _getTextButton(
                      widget.skipText ?? "SKIP",
                      (widget.skipCallback != null
                          ? widget.skipCallback
                          : _skip)),
                  opacity: (_counter < _total - 1) ? 1.0 : 0.0,
                ),
                Expanded(
                  child: Center(child: LayoutBuilder(
                    builder: (context, constraints) {
                      _bulletContainerWidth = constraints.maxWidth - 40.0;
                      return Container(
                        padding: const EdgeInsets.all(20.0),
                        child: (widget.showBullets
                            ? SingleChildScrollView(
                                physics: NeverScrollableScrollPhysics(),
                                scrollDirection: Axis.horizontal,
                                controller: _scrollController,
                                child: Row(
                                  mainAxisAlignment: MainAxisAlignment.center,
                                  children: <Widget>[
                                    for (int i = 0; i < _total; i++)
                                      Padding(
                                        padding: EdgeInsets.all(_bulletPadding),
                                        child: Container(
                                          height: _bulletSize,
                                          width: _bulletSize,
                                          decoration: BoxDecoration(
                                              shape: BoxShape.circle,
                                              color: (i == _counter
                                                  ? widget.activeBulletColor
                                                  : widget
                                                      .inactiveBulletColor)),
                                        ),
                                      )
                                  ],
                                ),
                              )
                            : Container()),
                      );
                    },
                  )),
                ),
                (_counter < _total - 1
                    ? _getTextButton(
                        (widget.nextText ?? "NEXT"),
                        _next,
                      )
                    : _getTextButton((widget.finishText ?? "FINISH"),
                        widget.finishCallback)),
              ],
            ),
          ),
        ],
      ),
    );
  }

  _getPage(index) {
    PageModel page = widget.pages[index];
    return Container(
      width: double.infinity,
      height: double.infinity,
      decoration: _boxDecoration.copyWith(
        color:
            widget.backgroundProvider == null ? page.color : Colors.transparent,
      ),
      child: page.child != null
          ? Center(
              child: page.doAnimateChild
                  ? AnimatedBoard(
                      animator: _animator,
                      child: page.child,
                    )
                  : page.child,
            )
          : Column(
              mainAxisAlignment: MainAxisAlignment.center,
              children: <Widget>[
                page.doAnimateImage
                    ? AnimatedBoard(
                        animator: _animator,
                        child: new Padding(
                          padding: new EdgeInsets.only(bottom: 25.0),
                          child: new Image.asset(page.imageAssetPath!,
                              width: 300.0, height: 300.0),
                        ),
                      )
                    : Image.asset(page.imageAssetPath!,
                        width: 300.0, height: 300.0),
                Padding(
                  padding: new EdgeInsets.only(
                      top: 10.0, bottom: 10.0, left: 30.0, right: 30.0),
                  child: new Text(
                    page.title!,
                    textAlign: TextAlign.center,
                    style: new TextStyle(
                      color: page.titleColor ?? Colors.white,
                      fontSize: 34.0,
                    ),
                  ),
                ),
                Padding(
                  padding: new EdgeInsets.only(
                      bottom: 75.0, left: 30.0, right: 30.0),
                  child: new Text(
                    page.body!,
                    textAlign: TextAlign.center,
                    style: new TextStyle(
                      color: page.bodyColor ?? Colors.white,
                      fontSize: 18.0,
                    ),
                  ),
                ),
              ],
            ),
    );
  }

  _getTextButton(_text, _onPress) {
    return TextButton(
      style: TextButton.styleFrom(
        padding: const EdgeInsets.all(20.0),
      ),
      child: Text(
        _text,
        style: TextStyle(color: widget.buttonColor),
      ),
      onPressed: _onPress,
    );
  }

  _goBack() {
    setState(() {
      _swipeDirection = SwipeDirection.LEFT_TO_RIGHT;
      _last = _counter;
      _counter--;
    });
    _animate();
  }

  _next() {
    setState(() {
      _swipeDirection = SwipeDirection.RIGHT_TO_LEFT;
      _last = _counter;
      _counter++;
    });
    _animate();
  }

  _skip() {
    setState(() {
      _swipeDirection = SwipeDirection.SKIP_TO_LAST;
      _last = _counter;
      _counter = _total - 1;
    });
    _animate();
  }

  _animate() {
    _animator.getController().forward(from: 0.0);

    double _bulletDimension = (_bulletPadding * 2) + (_bulletSize);
    double _scroll = _bulletDimension * _counter;
    double _maxScroll = _bulletDimension * _total - 1;
    if (_scroll > _bulletContainerWidth &&
        _swipeDirection == SwipeDirection.RIGHT_TO_LEFT) {
      double _scrollDistance =
          (((_scroll - _bulletContainerWidth) ~/ _bulletDimension) + 1) *
              _bulletDimension;
      _scrollController.animateTo(_scrollDistance,
          curve: Curves.easeIn, duration: Duration(milliseconds: 100));
    } else if (_scroll < (_maxScroll - _bulletContainerWidth) &&
        _swipeDirection == SwipeDirection.LEFT_TO_RIGHT) {
      _scrollController.animateTo(_scroll,
          curve: Curves.easeIn, duration: Duration(milliseconds: 100));
    } else if (_swipeDirection == SwipeDirection.SKIP_TO_LAST) {
      _scrollController.animateTo(_maxScroll,
          curve: Curves.easeIn, duration: Duration(milliseconds: 100));
    }
  }
}

class AnimatedBoard extends StatelessWidget {
  final Widget? child;
  final OverBoardAnimator animator;

  const AnimatedBoard({Key? key, required this.animator, required this.child})
      : super(key: key);
  @override
  Widget build(BuildContext context) {
    return Transform(
      transform: Matrix4.translationValues(
          0.0, 50.0 * (1.0 - animator.getAnimator().value), 0.0),
      child: new Padding(
        padding: new EdgeInsets.only(bottom: 25.0),
        child: child,
      ),
    );
  }
}


lib/src/page_model.dart
import 'package:flutter/cupertino.dart';

class PageModel {
  /// Background color of the page
  Color? color;

  /// Image path from asset to show in page
  String? imageAssetPath;

  /// Title text of the page
  String? title;

  /// Body text of the page
  String? body;

  /// Custom widget to pass as image in page
  Widget? child;

  /// To enable/disable child animation
  bool doAnimateChild = false;

  /// To enable/disable image animation
  bool doAnimateImage = false;

  /// Change color of title text
  Color? titleColor;

  /// Change color of body text
  Color? bodyColor;

  /// Create page model with image in show in on-boarding widget
  PageModel(
      {this.color,
      this.titleColor,
      this.bodyColor,
      required this.imageAssetPath,
      required this.title,
      required this.body,
      this.doAnimateImage = false});

  /// Create page model with custom child in show in on-boarding widget
  PageModel.withChild(
      {required this.child,
      required this.color,
      this.titleColor,
      this.bodyColor,
      this.doAnimateChild = false});
}


lib/src/overboard_animator.dart
import 'package:flutter/widgets.dart';

class OverBoardAnimator {
  late TickerProvider _vsync;
  late AnimationController _controller;
  late Animation _animation;

  //  static final OverBoardAnimator _singleton = new OverBoardAnimator._internal();

  // factory OverBoardAnimator(_vsync) {
  //   _singleton._vsync = _vsync;
  //   _singleton._controller = AnimationController(
  //       vsync: _vsync, duration: const Duration(milliseconds: 500));
  //   _singleton._animation = CurvedAnimation(parent: _singleton._controller, curve: Curves.easeIn);
  //   return _singleton;
  // }

  // OverBoardAnimator._internal();

  OverBoardAnimator(vsync) {
    this._vsync = vsync;
    _controller = AnimationController(
        vsync: _vsync, duration: const Duration(milliseconds: 500));
    _animation = CurvedAnimation(parent: _controller, curve: Curves.easeIn);
    print('creating new animator');
  }

  AnimationController getController() {
    return _controller;
  }

  Animation getAnimator() {
    return _animation;
  }

  dispose() {
    _controller.dispose();
  }
}


example/lib/main.dart
import 'package:flutter/material.dart';

import 'package:flutter_overboard/flutter_overboard.dart';

void main() => runApp(MyApp());

class MyApp extends StatelessWidget {
  @override
  Widget build(BuildContext context) {
    return MaterialApp(
      title: 'Flutter Overboard Demo',
      debugShowCheckedModeBanner: false,
      theme: ThemeData(
        primarySwatch: Colors.blue,
      ),
      home: MyHomePage(title: 'Flutter Overboard Home Page'),
    );
  }
}

class MyHomePage extends StatefulWidget {
  MyHomePage({Key? key, required this.title}) : super(key: key);

  final String title;

  @override
  _MyHomePageState createState() => _MyHomePageState();
}

class _MyHomePageState extends State<MyHomePage> with TickerProviderStateMixin {
  final GlobalKey<ScaffoldState> _globalKey = new GlobalKey<ScaffoldState>();

  @override
  void initState() {
    super.initState();
  }

  @override
  Widget build(BuildContext context) {
    return Scaffold(
      key: _globalKey,
      body: OverBoard(
        allowScroll: true,
        pages: pages,
        showBullets: true,
        inactiveBulletColor: Colors.blue,
        // backgroundProvider: NetworkImage('https://picsum.photos/720/1280'),
        skipCallback: () {
          ScaffoldMessenger.of(context).showSnackBar(SnackBar(
            content: Text("Skip clicked"),
          ));
        },
        finishCallback: () {
          ScaffoldMessenger.of(context).showSnackBar(SnackBar(
            content: Text("Finish clicked"),
          ));
        },
      ),
    );
  }

  final pages = [
    PageModel(
        color: const Color(0xFF0097A7),
        imageAssetPath: 'assets/01.png',
        title: 'Screen 1',
        body: 'Share your ideas with the team',
        doAnimateImage: true),
    PageModel(
        color: const Color(0xFF536DFE),
        imageAssetPath: 'assets/02.png',
        title: 'Screen 2',
        body: 'See the increase in productivity & output',
        doAnimateImage: true),
    PageModel(
        color: const Color(0xFF9B90BC),
        imageAssetPath: 'assets/03.png',
        title: 'Screen 3',
        body: 'Connect with the people from different places',
        doAnimateImage: true),
    PageModel.withChild(
        child: Padding(
          padding: EdgeInsets.only(bottom: 25.0),
          child: Image.asset('assets/02.png', width: 300.0, height: 300.0),
        ),
        color: const Color(0xFF5886d6),
        doAnimateChild: true)
  ];
}
